src/lib/datamodel/Deque.h

```diff
@@ -0,0 +1,186 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *  @file
+ *    basic double-ended queue.  Two intended use models:
+ *    1. used as a raw templated type
+ *     of an object or used as a base class
+ */
+
+#ifndef CHIP_DEQUE_H
+#define CHIP_DEQUE_H
+
+#include <functional>
+#include <stddef.h>
+
+namespace chip {
+
+/**
+ *  @class Deque
+ *
+ *  embed-able double-ended queue item with a recovery pointer
+ *
+ */
+template <class T>
+class Deque
+{
+
+public:
+    Deque(T * me) : mMe(me) { mNext = mPrev = this; };
+    ~Deque() { Remove(); };
+
+    Deque * Next() { return mNext; }
+    Deque * Prev() { return mPrev; }
+    T * Me() { return mMe; }
+
+    void Remove()
+    {
+        mNext->mPrev = mPrev;
+        mPrev->mNext = mNext;
+        mNext = mPrev = this;
+    }
+
+    /**
+     * @brief iterate over the list, starting from
+     *     this item, until fn() returns true.  Returns
+     *     item->mMe of the match or NULL if no match.
+     *  Mutation of the list during fn() should be avoided
+     */
+    T * Find(std::function<bool(T *)> fn)
```

Comment:
I know we are sensitive to code size and memory usage, what's the tradeoff here (should we be using std::function in a constrained environment)? Why not ordinary iterators?

Pigweed is working on container types in third_party/pigweed/repo/pw_containers and have similar goals. I see a list type there that we could make use of.

Reply:
std::function because I wanted to provide a function pointer+context in the most C++11-canonical way.  Expect an iterator would have similar (small) on-stack overhead.

Reply:
A type-erased construct like std::function is absolutely going to blow up the code-size on embedded targets.

Why not a lambda passed as a template parameter?

Reply:
a lambda was what I was aiming for, but found advice that std::function would allow more variations of callers